File: frontend/sabiops-frontend/src/services/analyticsCacheService.js
/**
 * Frontend Analytics Cache Service
 * Provides client-side caching for analytics data to improve performance
 */

class FrontendAnalyticsCacheService {
  constructor() {
    this.cache = new Map();
    this.cacheTimestamps = new Map();
    
    // Cache TTL settings (in milliseconds)
    this.TTL_SETTINGS = {
      'daily': 5 * 60 * 1000,    // 5 minutes for daily data
      'weekly': 15 * 60 * 1000,  // 15 minutes for weekly data
      'monthly': 30 * 60 * 1000, // 30 minutes for monthly data
      'yearly': 60 * 60 * 1000   // 1 hour for yearly data
    };
    
    // Maximum cache size to prevent memory issues
    this.MAX_CACHE_SIZE = 50;
    
    // Start periodic cleanup
    this.startCleanupInterval();
  }

  /**
   * Generate cache key for analytics data
   */
  getCacheKey(userId, analyticsType, timePeriod, additionalParams = {}) {
    const keyData = {
      userId,
      type: analyticsType,
      period: timePeriod,
      ...additionalParams
    };
    return JSON.stringify(keyData);
  }

  /**
   * Get cached data if it exists and is not expired
   */
  getCachedData(cacheKey) {
    try {
      if (!this.cache.has(cacheKey)) {
        return null;
      }

      const timestamp = this.cacheTimestamps.get(cacheKey);
      const now = Date.now();
      
      // Extract time period from cache key to determine TTL
      const keyData = JSON.parse(cacheKey);
      const ttl = this.TTL_SETTINGS[keyData.period] || this.TTL_SETTINGS.monthly;
      
      if (now - timestamp > ttl) {
        // Cache expired
        this.cache.delete(cacheKey);
        this.cacheTimestamps.delete(cacheKey);
        return null;
      }

      const cachedData = this.cache.get(cacheKey);
      console.log(`[Analytics Cache] Cache hit for: ${keyData.type} (${keyData.period})`);
      return cachedData;
      
    } catch (error) {
      console.error('[Analytics Cache] Error retrieving cached data:', error);
      return null;
    }
  }

  /**
   * Store data in cache
   */
  setCachedData(cacheKey, data, timePeriod) {
    try {
      // Check cache size and clean up if necessary
      if (this.cache.size >= this.MAX_CACHE_SIZE) {
        this.cleanupOldestEntries(10); // Remove 10 oldest entries
      }

      this.cache.set(cacheKey, data);
      this.cacheTimestamps.set(cacheKey, Date.now());
      
      const keyData = JSON.parse(cacheKey);
      console.log(`[Analytics Cache] Data cached: ${keyData.type} (${keyData.period})`);
      return true;
      
    } catch (error) {
      console.error('[Analytics Cache] Error caching data:', error);
      return false;
    }
  }

  /**
   * Invalidate cache for specific user
   */
  invalidateUserCache(userId) {
    try {
      let removedCount = 0;
      
      for (const [cacheKey] of this.cache) {
        try {
          const keyData = JSON.parse(cacheKey);
          if (keyData.userId === userId) {
            this.cache.delete(cacheKey);
            this.cacheTimestamps.delete(cacheKey);
            removedCount++;
          }
        } catch (error) {
          // Invalid cache key format, remove it
          this.cache.delete(cacheKey);
          this.cacheTimestamps.delete(cacheKey);
          removedCount++;
        }
      }
      
      console.log(`[Analytics Cache] Invalidated ${removedCount} entries for user ${userId}`);
      return removedCount;
      
    } catch (error) {
      console.error('[Analytics Cache] Error invalidating user cache:', error);
      return 0;
    }
  }

  /**
   * Clear all expired cache entries
   */
  clearExpiredCache() {
    try {
      const now = Date.now();
      let removedCount = 0;
      
      for (const [cacheKey, timestamp] of this.cacheTimestamps) {
        try {
          const keyData = JSON.parse(cacheKey);
          const ttl = this.TTL_SETTINGS[keyData.period] || this.TTL_SETTINGS.monthly;
          
          if (now - timestamp > ttl) {
            this.cache.delete(cacheKey);
            this.cacheTimestamps.delete(cacheKey);
            removedCount++;
          }
        } catch (error) {
          // Invalid cache key, remove it
          this.cache.delete(cacheKey);
          this.cacheTimestamps.delete(cacheKey);
          removedCount++;
        }
      }
      
      if (removedCount > 0) {
        console.log(`[Analytics Cache] Cleared ${removedCount} expired entries`);
      }
      
      return removedCount;
      
    } catch (error) {
      console.error('[Analytics Cache] Error clearing expired cache:', error);
      return 0;
    }
  }

  /**
   * Remove oldest cache entries
   */
  cleanupOldestEntries(count) {
    try {
      // Convert to array and sort by timestamp
      const entries = Array.from(this.cacheTimestamps.entries())
        .sort((a, b) => a[1] - b[1]) // Sort by timestamp (oldest first)
        .slice(0, count); // Take the oldest entries
      
      for (const [cacheKey] of entries) {
        this.cache.delete(cacheKey);
        this.cacheTimestamps.delete(cacheKey);
      }
      
      console.log(`[Analytics Cache] Cleaned up ${entries.length} oldest entries`);
      return entries.length;
      
    } catch (error) {
      console.error('[Analytics Cache] Error cleaning up oldest entries:', error);
      return 0;
    }
  }

  /**
   * Get cache statistics
   */
  getCacheStats() {
    try {
      const now = Date.now();
      let expiredCount = 0;
      
      for (const [cacheKey, timestamp] of this.cacheTimestamps) {
        try {
          const keyData = JSON.parse(cacheKey);
          const ttl = this.TTL_SETTINGS[keyData.period] || this.TTL_SETTINGS.monthly;
          
          if (now - timestamp > ttl) {
            expiredCount++;
          }
        } catch (error) {
          expiredCount++;
        }
      }
      
      return {
        totalEntries: this.cache.size,
        activeEntries: this.cache.size - expiredCount,
        expiredEntries: expiredCount,
        maxCacheSize: this.MAX_CACHE_SIZE,
        cacheUtilization: (this.cache.size / this.MAX_CACHE_SIZE * 100).toFixed(1) + '%'
      };
      
    } catch (error) {
      console.error('[Analytics Cache] Error getting cache stats:', error);
      return {
        totalEntries: 0,
        activeEntries: 0,
        expiredEntries: 0,
        error: error.message
      };
    }
  }

  /**
   * Clear all cache
   */
  clearAllCache() {
    try {
      const clearedCount = this.cache.size;
      this.cache.clear();
      this.cacheTimestamps.clear();
      
      console.log(`[Analytics Cache] Cleared all cache (${clearedCount} entries)`);
      return clearedCount;
      
    } catch (error) {
      console.error('[Analytics Cache] Error clearing all cache:', error);
      return 0;
    }
  }

  /**
   * Start periodic cleanup of expired entries
   */
  startCleanupInterval() {
    // Clean up expired entries every 10 minutes
    setInterval(() => {
      this.clearExpiredCache();
    }, 10 * 60 * 1000);
    
    console.log('[Analytics Cache] Started periodic cleanup interval');
  }

  /**
   * Preload analytics data for better user experience
   */
  async preloadAnalyticsData(userId, currentPeriod, apiCall) {
    try {
      const periods = ['daily', 'weekly', 'monthly', 'yearly'];
      const otherPeriods = periods.filter(p => p !== currentPeriod);
      
      // Preload data for other time periods in the background
      for (const period of otherPeriods.slice(0, 2)) { // Only preload 2 other periods
        const cacheKey = this.getCacheKey(userId, 'business_analytics', period);
        
        if (!this.getCachedData(cacheKey)) {
          try {
            // Make API call in background
            setTimeout(async () => {
              try {
                const data = await apiCall(period);
                this.setCachedData(cacheKey, data, period);
              } catch (error) {
                console.warn(`[Analytics Cache] Preload failed for ${period}:`, error);
              }
            }, 1000); // Delay to not interfere with current request
          } catch (error) {
            console.warn(`[Analytics Cache] Preload setup failed for ${period}:`, error);
          }
        }
      }
      
    } catch (error) {
      console.error('[Analytics Cache] Error in preload setup:', error);
    }
  }
}

// Create and export singleton instance
const frontendAnalyticsCache = new FrontendAnalyticsCacheService();

export default frontendAnalyticsCache;

File: backend/sabiops-backend/src/services/analytics_cache_service.py
"""
Analytics Cache Service
Provides caching functionality for analytics data to improve performance
"""

import json
import hashlib
from datetime import datetime, timedelta, timezone
from typing import Dict, Optional, Any
import logging

logger = logging.getLogger(__name__)

class AnalyticsCacheService:
    """Service for caching analytics data to improve performance"""
    
    def __init__(self):
        # In-memory cache for development
        # In production, you would use Redis or similar
        self._cache = {}
        self._cache_ttl = {}
        
        # Cache TTL settings (in seconds)
        self.TTL_SETTINGS = {
            'daily': 300,    # 5 minutes for daily data
            'weekly': 900,   # 15 minutes for weekly data
            'monthly': 1800, # 30 minutes for monthly data
            'yearly': 3600   # 1 hour for yearly data
        }
    
    def get_cache_key(self, user_id: str, analytics_type: str, time_period: str, **kwargs) -> str:
        """Generate a unique cache key for analytics data"""
        key_data = {
            'user_id': user_id,
            'type': analytics_type,
            'period': time_period,
            **kwargs
        }
        key_string = json.dumps(key_data, sort_keys=True)
        return hashlib.md5(key_string.encode()).hexdigest()
    
    def get_cached_data(self, cache_key: str) -> Optional[Dict[str, Any]]:
        """Retrieve cached data if it exists and is not expired"""
        try:
            if cache_key not in self._cache:
                return None
            
            # Check if cache has expired
            if cache_key in self._cache_ttl:
                expiry_time = self._cache_ttl[cache_key]
                if datetime.now(timezone.utc) > expiry_time:
                    # Cache expired, remove it
                    del self._cache[cache_key]
                    del self._cache_ttl[cache_key]
                    return None
            
            cached_data = self._cache[cache_key]
            logger.info(f"Cache hit for key: {cache_key}")
            return cached_data
            
        except Exception as e:
            logger.error(f"Error retrieving cached data: {str(e)}")
            return None
    
    def set_cached_data(self, cache_key: str, data: Dict[str, Any], time_period: str) -> bool:
        """Store data in cache with appropriate TTL"""
        try:
            ttl_seconds = self.TTL_SETTINGS.get(time_period, 1800)  # Default 30 minutes
            expiry_time = datetime.now(timezone.utc) + timedelta(seconds=ttl_seconds)
            
            self._cache[cache_key] = data
            self._cache_ttl[cache_key] = expiry_time
            
            logger.info(f"Data cached with key: {cache_key}, TTL: {ttl_seconds}s")
            return True
            
        except Exception as e:
            logger.error(f"Error caching data: {str(e)}")
            return False
    
    def invalidate_user_cache(self, user_id: str) -> int:
        """Invalidate all cached data for a specific user"""
        try:
            keys_to_remove = []
            for cache_key in self._cache.keys():
                # Check if this cache key belongs to the user
                if user_id in cache_key:
                    keys_to_remove.append(cache_key)
            
            # Remove the keys
            for key in keys_to_remove:
                if key in self._cache:
                    del self._cache[key]
                if key in self._cache_ttl:
                    del self._cache_ttl[key]
            
            logger.info(f"Invalidated {len(keys_to_remove)} cache entries for user {user_id}")
            return len(keys_to_remove)
            
        except Exception as e:
            logger.error(f"Error invalidating user cache: {str(e)}")
            return 0
    
    def clear_expired_cache(self) -> int:
        """Remove all expired cache entries"""
        try:
            current_time = datetime.now(timezone.utc)
            expired_keys = []
            
            for cache_key, expiry_time in self._cache_ttl.items():
                if current_time > expiry_time:
                    expired_keys.append(cache_key)
            
            # Remove expired entries
            for key in expired_keys:
                if key in self._cache:
                    del self._cache[key]
                if key in self._cache_ttl:
                    del self._cache_ttl[key]
            
            logger.info(f"Cleared {len(expired_keys)} expired cache entries")
            return len(expired_keys)
            
        except Exception as e:
            logger.error(f"Error clearing expired cache: {str(e)}")
            return 0
    
    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics for monitoring"""
        try:
            current_time = datetime.now(timezone.utc)
            total_entries = len(self._cache)
            expired_entries = 0
            
            for expiry_time in self._cache_ttl.values():
                if current_time > expiry_time:
                    expired_entries += 1
            
            return {
                'total_entries': total_entries,
                'active_entries': total_entries - expired_entries,
                'expired_entries': expired_entries,
                'cache_hit_ratio': getattr(self, '_hit_ratio', 0.0),
                'last_cleanup': getattr(self, '_last_cleanup', None)
            }
            
        except Exception as e:
            logger.error(f"Error getting cache stats: {str(e)}")
            return {
                'total_entries': 0,
                'active_entries': 0,
                'expired_entries': 0,
                'error': str(e)
            }

# Global cache service instance
analytics_cache = AnalyticsCacheService()